### PR TITLE
Convert eigenfunctions

### PR DESCRIPTION
**Pull request by dirmeier**
## Summary

Hello all, 
this PR converts `eigenvalues_sym` and `eigenvectors_sym` to use arena types (sorry for the delay, busy weeks).

Cheers,
Simon

## Tests

`eigenvalues_sym` is already tested, but I noticed that `eigenvectors_sym` does not have a corresponding test in `math/rev/fun` and still needs some implementation. I'm not sure how to test this though. Does anyone have a suggestion maybe? Thanks!

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Simon Dirmeier

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Nice!! I'll give this a look either today or Monday

File: stan/math/rev/fun/eigenvectors_sym.hpp [hunk 1]
```diff
@@ -14,86 +14,42 @@
...
+  reverse_pass_callback([arena_m, eigenvals, eigenvecs]() mutable {
+    const int p = arena_m.val().cols();
+    auto eigenvecs_a = to_arena(eigenvecs.val());
+    auto eigenvecs_a_adj = to_arena(eigenvecs.adj());
+
+    matrix_d f(p, p);
+    for (int i = 0; i < p; ++i)
+      for (int j = 0; j < p; ++j)
+        f.coeffRef(j, i)
+            = (i != j ? 1 / (eigenvals.coeff(i) - eigenvals.coeff(j)) : 0);
```
Comment (SteveBronder) [inline stan/math/rev/fun/eigenvectors_sym.hpp:45]: We tend to use i for the rows and j for the columns. And I think you can make this

```cpp
    Eigen::MatrixXd f = (1 / (eigenvals.rowwise().replicate(p).transpose() - eigenvals.rowwise().replicate(p)).array());
    f.diagonal().setZero();
```

We could do more clever things with `selfAdjointView` but idk how to do elementwise multiplication with a `selfAdjointView` which we would need below

**Review by SteveBronder — CHANGES_REQUESTED** [review]
This looks very nice!

You should also add tests for the `var<Matrix>` type in `mix/fun/eigenvalues_sym` and `mix/fun/eigenvectors_sym` tests. For that you should only need to change the lambda `f` in each of those tests so it calls Stan's `multiply` instead of `*` on this line 

```cpp
    auto a = stan::math::multiply((y + y.transpose()), 0.5).eval();
```

Then you should be able to just take all the places that are like 

```cpp
  stan::test::expect_ad(tols, f, a22);
```

and put this right below them

```cpp
  stan::test::expect_ad_matvar(tols, f, a22);
```

**Comment by dirmeier** [comment]
Ciao @SteveBronder , thanks for the review and the tips. I've added the tests as you suggested and fixed the other things. 

When I'm changing the nonzero size check to `unlikely(m.size() == 0`, the `rev` and `mix` tests start failing though (when the matrix size is 0), so I needed to remove them. Also I think it leads to [failing expression tests](https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2476/5/pipeline#step-232-log-101). Is that possible?

Thanks in advance!

Cheers,
Simon

File: stan/math/rev/fun/eigenvectors_sym.hpp [hunk 2]
```diff
@@ -14,86 +14,42 @@
...
-  res.vi()
-      = Eigen::Map<matrix_vi>(baseVari->vari_ref_v_, res.rows(), res.cols());
-  return res;
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline auto eigenvectors_sym(const T& m) {
+  check_nonzero_size("eigenvalues_sym", "m", m);
+  check_symmetric("eigenvalues_sym", "m", m);
+
+  using return_t = return_var_matrix_t<T>;
+  auto arena_m = to_arena(m);
```
[the remark this answers is not in the document]
Comment (dirmeier) [reply stan/math/rev/fun/eigenvectors_sym.hpp]: Alright. That's tidier indeed..

**Comment by SteveBronder** [comment]
Also I think it leads to failing expression tests. Is that possible?

Oh yes sorry for that you just need to do 

```cpp
if (unlikely(m.size() == 0)) {
  return return_t(m);
}
```

What's happening with an error like the below is that we return back two different types (keeping in mind the input argument is not going to be the same as the type in `return_t`

```cpp
./stan/math/rev/fun/eigenvalues_sym.hpp:42:3: error: 'auto' in return type deduced as 'Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>' here but deduced as 'Eigen::CwiseUnaryOp<stan::test::counterOp<stan::math::var_value<double, void> >, const Eigen::Block<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, -1, -1, false> >' in earlier return statement

  return return_t(eigenvals);
  ^

test/expressions/tests34_test.cpp:10592:45: note: in instantiation of function template specialization 'stan::math::eigenvalues_sym<Eigen::CwiseUnaryOp<stan::test::counterOp<stan::math::var_value<double, void> >, const Eigen::Block<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, -1, -1, false> >, nullptr>' requested here

auto result3 = stan::math::eval(stan::math::eigenvalues_sym(matrix0_expr1));
```

So the early exit returns back an `Eigen::Block` but the success returns back a `Matrix`. This makes `auto` confused about the return type since two paths return two different types and so it throws an error. We can fix that by just setting the early returns type to `return_t()`

**Comment by dirmeier** [comment]
My bad, should have had a closer look. Thanks!

**Review by SteveBronder — APPROVED** [review]
Looks good!! Feel free to click the button to merge

**Pull request merged**